services/src/functional-tests/src/test/groovy/org/openkilda/functionaltests/spec/flows/ProtectedPathV2Spec.groovy

```diff
@@ -1183,6 +1184,68 @@ class ProtectedPathV2Spec extends HealthCheckSpecification {
         northbound.deleteLinkProps(northbound.getAllLinkProps())
     }
 
+    @Tidy
+    def "Protected path is created in different POP even if this path is not preferable"(){
+        given: "Not neighboring switch pair with three diverse paths at least"
+        def swPair = topologyHelper.getAllNotNeighboringSwitchPairs().find { swPair ->
+            swPair.paths.findAll { pathHelper.getInvolvedSwitches(it).size() == 3 }.unique(false) { a, b ->
+                def p1 = pathHelper.getInvolvedSwitches(a)[1..-2]*.dpId
+                def p2 = pathHelper.getInvolvedSwitches(b)[1..-2]*.dpId
+                p1.intersect(p2) == [] ? 1 : 0
+            }.size() >= 3
+        } ?: assumeTrue("No suiting switches found", false)
+
+        //find out all paths with 3 switches
+        def allPaths = swPair.paths.findAll { pathHelper.getInvolvedSwitches(it).size() == 3 }
+        def allPathCandidates = allPaths.unique(false) { a, b ->
+            a.intersect(b) == [] ? 1 : 0
+        }
+        assumeTrue("Not enough amount of diverse paths", allPathCandidates.size() >= 3)
+        //select paths for further manipulations
+        def mainPath1 = allPathCandidates.first()
+        def mainPath2 = allPathCandidates.find { it != mainPath1 }
+        def protectedPath = allPathCandidates.find { it != mainPath1 && it != mainPath2}
+        def involvedSwP1 = pathHelper.getInvolvedSwitches(mainPath1)*.dpId
+        def involvedSwP2 = pathHelper.getInvolvedSwitches(mainPath2)*.dpId
+        def involvedSwProtected = pathHelper.getInvolvedSwitches(protectedPath)*.dpId
+
+        and: "Src, dst and transit switches belongs to different POPs(src:1, dst:4, tr1/tr2:2, tr3:3)"
+        database.setSwitchPop(swPair.src.dpId, "1")
+        [involvedSwP1[1], involvedSwP2[1]].each { swId -> database.setSwitchPop(swId, "2") }
+        database.setSwitchPop(swPair.dst.dpId, "3")
+        database.setSwitchPop(involvedSwProtected[1], "4")
+
+        and: "Path which contains tr3 is non preferable"
+        allPaths.findAll { involvedSwProtected[1] in pathHelper.getInvolvedSwitches(it)*.dpId }.each {
+            pathHelper.getInvolvedIsls(it).each {
+                northbound.updateLinkProps([islUtils.toLinkProps(it,
+                        ["cost": (database.getIslCost(it) + 10).toString()])])
+            }
+        }
+
+        when: "Create a protected flow"
+        def flow = flowHelperV2.randomFlow(swPair)
+        flow.allocateProtectedPath = true
+        flowHelperV2.addFlow(flow)
+
+        then: "Main path is built through the preferable path(tr1 or tr2)"
+        def flowPaths = northbound.getFlowPath(flow.flowId)
+        def realFlowPathInvolvedSwitches = pathHelper.getInvolvedSwitches(pathHelper.convert(flowPaths))*.dpId
+        realFlowPathInvolvedSwitches == involvedSwP1 || realFlowPathInvolvedSwitches == involvedSwP2
+
+        and: "Protected path is built through the non preferable path(tr3)"
```

Comment:
can you elaborate here please? I mean the reason why it picked 'tr3' path. It is really hard to follow all this pop logic without any pictures before eyes and with very brief comments. I mean I get it right now, but it will take me too much time to refresh the memory in a month or two. And the test must be a good helper here

Reply:
left the comment